compiler/dcalc/from_scopelang.ml

```diff
@@ -307,9 +307,121 @@ let rec translate_expr (ctx : 'm ctx) (e : 'm Scopelang.Ast.expr) :
     let arg_struct =
       Expr.estruct sc_sig.scope_sig_input_struct field_map (mark_tany m pos)
     in
-    Expr.eapp
-      (Expr.evar sc_sig.scope_sig_scope_var (mark_tany m pos))
-      [arg_struct] m
+    let called_func =
+      tag_with_log_entry
+        (Expr.evar sc_sig.scope_sig_scope_var (mark_tany m pos))
+        BeginCall
+        [ScopeName.get_info scope; Marked.mark (Expr.pos e) "direct"]
+    in
+    let single_arg =
+      tag_with_log_entry arg_struct
+        (VarDef (TStruct sc_sig.scope_sig_input_struct))
+        [
+          ScopeName.get_info scope;
+          Marked.mark (Expr.pos e) "direct";
+          Marked.mark (Expr.pos e) "input";
+        ]
+    in
+    let direct_output_info =
+      [
+        ScopeName.get_info scope;
+        Marked.mark (Expr.pos e) "direct";
+        Marked.mark (Expr.pos e) "output";
+      ]
+    in
+    let calling_expr = Expr.eapp called_func [single_arg] m in
+    (* For the purposes of log parsing explained in Runtime.EventParser, we need
+       to wrap this function call in a flurry of log tags. Specifically, we are
+       mascarading this scope call as a function call. In a normal function
+       call, the log parser expects the output of the function to be defined as
+       a default, hence the production of the output should yield a
+       PosRecordIfTrueBool (which is not the case here). To remedy this absence
+       we fabricate a fake PosRecordIfTrueBool attached to a silent let binding
+       to "true" before returning the output value.
+
+       But this is not sufficient. Indeed for the tricky case of
+       [tests/test_scope/scope_call3.catala_en], when a scope returns a
+       function, because we insert loggins calls at the call site of the
+       function and not during its definition, then we're missing the call log
+       instructions of the function returned. To avoid this trap, we need to
+       rebind the resulting scope output struct by eta-expanding the functions
+       to insert logging instructions*)
+    let result_var = Var.make "result" in
+    let result_eta_expanded_var = Var.make "result" in
+    let result_eta_expanded =
+      Expr.estruct sc_sig.scope_sig_output_struct
```

Comment:
When going through the tedious task of manually building AST chunks, I find it helps a lot to write them in comments above as (pseudo-)syntax, both to understand what I am doing and to help reading the code.